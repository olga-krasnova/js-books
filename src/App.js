import React from 'react';
import ApiService from './services/ApiService';
import BookStoreService from './services/BookstoreService';

const apiService = new ApiService();

const getBooks = async () => {
  const bookStoreService = new BookStoreService(apiService);
  return await bookStoreService.fetchBooks().then(res => res);
};

const App = () => {
  console.log(getBooks());
  return <div className="App">Hello</div>;
};

export default App;
