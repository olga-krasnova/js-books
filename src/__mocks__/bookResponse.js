export default [
  {
    title: 'Effective JavaScript',
    subtitle: '68 Specific Ways to Harness the Power of JavaScript',
    isbn13: '9780321812186',
    price: '$21.99',
    image: 'https://itbook.store/img/books/9780321812186.png',
    url: 'https://itbook.store/books/9780321812186',
  },
  {
    title: 'Learning JavaScript',
    subtitle: 'A Hands-On Guide to the Fundamentals of Modern JavaScript',
    isbn13: '9780321832740',
    price: '$8.99',
    image: 'https://itbook.store/img/books/9780321832740.png',
    url: 'https://itbook.store/books/9780321832740',
  },
  {
    title: 'JavaScript The Good Parts',
    subtitle: 'Unearthing the Excellence in JavaScript',
    isbn13: '9780596517748',
    price: '$6.68',
    image: 'https://itbook.store/img/books/9780596517748.png',
    url: 'https://itbook.store/books/9780596517748',
  },
  {
    title: 'Head First JavaScript',
    subtitle: "A Learner's Companion to JavaScript",
    isbn13: '9780596527747',
    price: '$5.99',
    image: 'https://itbook.store/img/books/9780596527747.png',
    url: 'https://itbook.store/books/9780596527747',
  },
  {
    title: 'JavaScript Creativity',
    subtitle: 'Exploring the Modern Capabilities of JavaScript and HTML5',
    isbn13: '9781430259442',
    price: '$37.95',
    image: 'https://itbook.store/img/books/9781430259442.png',
    url: 'https://itbook.store/books/9781430259442',
  },
  {
    title: 'JavaScript for PHP Developers',
    subtitle: 'A Concise Guide to Mastering JavaScript',
    isbn13: '9781449320195',
    price: '$10.50',
    image: 'https://itbook.store/img/books/9781449320195.png',
    url: 'https://itbook.store/books/9781449320195',
  },
  {
    title: 'Learning JavaScript Design Patterns',
    subtitle: "A JavaScript and jQuery Developer's Guide",
    isbn13: '9781449331818',
    price: '$31.21',
    image: 'https://itbook.store/img/books/9781449331818.png',
    url: 'https://itbook.store/books/9781449331818',
  },
  {
    title: 'The Problem with Native JavaScript APIs',
    subtitle: 'Why reliance on native JavaScript APIs leads to disadvantages',
    isbn13: '9781449339951',
    price: '$0.00',
    image: 'https://itbook.store/img/books/9781449339951.png',
    url: 'https://itbook.store/books/9781449339951',
  },
  {
    title: 'JavaScript Enlightenment',
    subtitle: 'From Library User to JavaScript Developer',
    isbn13: '9781449342883',
    price: '$9.93',
    image: 'https://itbook.store/img/books/9781449342883.png',
    url: 'https://itbook.store/books/9781449342883',
  },
  {
    title: 'JavaScript Testing with Jasmine',
    subtitle: 'JavaScript Behavior-Driven Development',
    isbn13: '9781449356378',
    price: '$10.25',
    image: 'https://itbook.store/img/books/9781449356378.png',
    url: 'https://itbook.store/books/9781449356378',
  },
];
