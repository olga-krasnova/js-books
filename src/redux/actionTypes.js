export const FETCH_BOOKS_SUCCESS = 'FETCH_BOOKS_SUCCESS';
export const FETCH_BOOKS_REQUEST = 'FETCH_BOOKS_REQUEST';
export const FETCH_BOOKS_FAILURE = 'FETCH_BOOKS_FAILURE';

export const BOOK_ADDED_TO_CART = 'BOOK_ADDED_TO_CART';
export const BOOK_REMOVED_FROM_CART = 'BOOK_REMOVED_FROM_CART';
export const ALL_BOOKS_REMOVED_FROM_CART = 'ALL_BOOKS_REMOVED_FROM_CART';
