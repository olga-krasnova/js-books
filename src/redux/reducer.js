const reducer = (state, action) => {
  if (state === undefined) {
    return {
      books: [],
      loading: true,
      error: null,
    };
  }

  switch (action.type) {
    case 'FETCH_BOOKS_REQUEST':
      return {
        books: [],
        loading: true,
        error: null,
      };

    default:
      return state;
  }
};

export default reducer;
