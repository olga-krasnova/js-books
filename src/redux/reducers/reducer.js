import updateBookList from './BookList';
import updateShoppingCart from './ShoppingCart';

const reducer = (state, action) => {
  return {
    bookList: updateBookList(state, action),
    shoppingCart: updateShoppingCart(state, action)
  };
};

export default reducer;
