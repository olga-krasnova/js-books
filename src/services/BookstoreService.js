const apiUrl = '/search/javascript';

export default class BookstoreService {
  constructor(apiService) {
    this.apiService = apiService;
  }

  async fetchBooks() {
    const result = await this.apiService
      .get(apiUrl)
      .then(response => response);

    return result.data.books;
  };
}
