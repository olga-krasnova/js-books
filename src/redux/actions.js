import * as types from './actionTypes';

const booksLoaded = (newBooks) => {
  return {
    type: types.FETCH_BOOKS_SUCCESS,
    payload: newBooks
  }
};

const booksRequested = () => {
  return {
    type: types.FETCH_BOOKS_REQUEST
  }
};

const booksError = (error) => {
  return {
    type: types.FETCH_BOOKS_FAILURE,
    payload: error
  }
};

export const bookAddedToCart = (bookId) => {
  return {
    type: types.BOOK_ADDED_TO_CART,
    payload: bookId
  }
};

export const bookRemovedFromCart = (bookId) => {
  return {
    type: types.BOOK_REMOVED_FROM_CART,
    payload: bookId
  }
};

export const allBooksRemovedFromCart = (bookId) => {
  return {
    type: types.ALL_BOOKS_REMOVED_FROM_CART,
    payload: bookId
  }
};

export const fetchBooks = (bookstoreService, dispatch) => () => {
  dispatch(booksRequested());
  bookstoreService.fetchBooks()
    .then((data) => dispatch(booksLoaded(data)))
    .catch((error) => dispatch(booksError(error)));
};
