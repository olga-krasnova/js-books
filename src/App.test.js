import React from 'react';
import { applyMiddleware, createStore } from 'redux';
import thunk from 'redux-thunk';
import { Provider } from 'react-redux';
import { createMemoryHistory } from 'history';
import { Router } from 'react-router-dom';
import { render, wait } from '@testing-library/react';
import reducer from './redux/reducers/reducer';
import App from './App';
import BookstoreService from './services/BookstoreService';
import BookstoreServiceContext from './components/contexts/BookstoreServiceContext';
import bookResponse from './__mocks__/bookResponse';
import mockAxios from './__mocks__/axios';

const bookstoreService = new BookstoreService(mockAxios);
let mockFetchBooks;

const initialState = {
  bookList: {
    books: [],
    isLoading: false,
    error: false,
  },
  shoppingCart: {
    cartItems: [],
    orderTotal: 0,
    itemsTotal: 0,
  },
};

function renderWithRouter(
  ui,
  {
    route = '/',
    history = createMemoryHistory({ initialEntries: [route] }),
    store = createStore(
      reducer,
      initialState,
      applyMiddleware(thunk.withExtraArgument({ fetchBooks: mockFetchBooks })),
    ),
    ...options
  } = {},
) {
  return {
    ...render(
      <Provider store={store}>
        <BookstoreServiceContext.Provider value={bookstoreService}>
          <Router history={history}>{ui}</Router>
        </BookstoreServiceContext.Provider>
      </Provider>,
      options,
    ),
    history,
    store,
  };
}

describe('App Router test', () => {
  test('should show BookList on home page', async () => {
    // Arrange
    const route = '/';

    mockFetchBooks = jest.fn();
    const mockApiServiceGet = jest.fn(
      () =>
        new Promise(resolve => {
          resolve({ data: { books: bookResponse } });
        }),
    );

    mockAxios.get.mockImplementation(mockApiServiceGet);

    // Act
    const { getByTestId } = await renderWithRouter(<App />, { route });
    const bookList = getByTestId('test-book-list');

    // Assert
    await wait(() => {
      expect(bookList).toMatchSnapshot();
    });
  });

  test('should show ShoppingCartTable on /cart', () => {
    // Arrange
    const route = '/cart';

    // Act
    const { getByTestId } = renderWithRouter(<App />, { route });
    const cartTable = getByTestId('test-cart-table');

    // Assert
    expect(cartTable).toMatchSnapshot();
  });
});
