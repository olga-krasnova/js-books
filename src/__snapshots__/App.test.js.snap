// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Router test should show BookList on home page 1`] = `
<ul
  data-testid="test-book-list"
>
  <li>
    <div
      class="sc-gzVnrw hnuYoW"
    >
      <div
        class="sc-htoDjs cEeFSy"
      >
        <img
          alt="Book cover"
          src="https://itbook.store/img/books/9780321812186.png"
        />
      </div>
      <div
        class="sc-dnqmqq bhqlKj"
      >
        <span
          class="sc-iwsKbI fJtlAM"
        >
          Effective JavaScript
        </span>
        <div />
        <div
          class="sc-gZMcBi cZDwvA"
        >
          $21.99
        </div>
        <button
          class="sc-bZQynM dlVnGl"
          type="button"
        >
          Add to cart
        </button>
      </div>
    </div>
  </li>
  <li>
    <div
      class="sc-gzVnrw hnuYoW"
    >
      <div
        class="sc-htoDjs cEeFSy"
      >
        <img
          alt="Book cover"
          src="https://itbook.store/img/books/9780321832740.png"
        />
      </div>
      <div
        class="sc-dnqmqq bhqlKj"
      >
        <span
          class="sc-iwsKbI fJtlAM"
        >
          Learning JavaScript
        </span>
        <div />
        <div
          class="sc-gZMcBi cZDwvA"
        >
          $8.99
        </div>
        <button
          class="sc-bZQynM dlVnGl"
          type="button"
        >
          Add to cart
        </button>
      </div>
    </div>
  </li>
  <li>
    <div
      class="sc-gzVnrw hnuYoW"
    >
      <div
        class="sc-htoDjs cEeFSy"
      >
        <img
          alt="Book cover"
          src="https://itbook.store/img/books/9780596517748.png"
        />
      </div>
      <div
        class="sc-dnqmqq bhqlKj"
      >
        <span
          class="sc-iwsKbI fJtlAM"
        >
          JavaScript The Good Parts
        </span>
        <div />
        <div
          class="sc-gZMcBi cZDwvA"
        >
          $6.68
        </div>
        <button
          class="sc-bZQynM dlVnGl"
          type="button"
        >
          Add to cart
        </button>
      </div>
    </div>
  </li>
  <li>
    <div
      class="sc-gzVnrw hnuYoW"
    >
      <div
        class="sc-htoDjs cEeFSy"
      >
        <img
          alt="Book cover"
          src="https://itbook.store/img/books/9780596527747.png"
        />
      </div>
      <div
        class="sc-dnqmqq bhqlKj"
      >
        <span
          class="sc-iwsKbI fJtlAM"
        >
          Head First JavaScript
        </span>
        <div />
        <div
          class="sc-gZMcBi cZDwvA"
        >
          $5.99
        </div>
        <button
          class="sc-bZQynM dlVnGl"
          type="button"
        >
          Add to cart
        </button>
      </div>
    </div>
  </li>
  <li>
    <div
      class="sc-gzVnrw hnuYoW"
    >
      <div
        class="sc-htoDjs cEeFSy"
      >
        <img
          alt="Book cover"
          src="https://itbook.store/img/books/9781430259442.png"
        />
      </div>
      <div
        class="sc-dnqmqq bhqlKj"
      >
        <span
          class="sc-iwsKbI fJtlAM"
        >
          JavaScript Creativity
        </span>
        <div />
        <div
          class="sc-gZMcBi cZDwvA"
        >
          $37.95
        </div>
        <button
          class="sc-bZQynM dlVnGl"
          type="button"
        >
          Add to cart
        </button>
      </div>
    </div>
  </li>
  <li>
    <div
      class="sc-gzVnrw hnuYoW"
    >
      <div
        class="sc-htoDjs cEeFSy"
      >
        <img
          alt="Book cover"
          src="https://itbook.store/img/books/9781449320195.png"
        />
      </div>
      <div
        class="sc-dnqmqq bhqlKj"
      >
        <span
          class="sc-iwsKbI fJtlAM"
        >
          JavaScript for PHP Developers
        </span>
        <div />
        <div
          class="sc-gZMcBi cZDwvA"
        >
          $10.50
        </div>
        <button
          class="sc-bZQynM dlVnGl"
          type="button"
        >
          Add to cart
        </button>
      </div>
    </div>
  </li>
  <li>
    <div
      class="sc-gzVnrw hnuYoW"
    >
      <div
        class="sc-htoDjs cEeFSy"
      >
        <img
          alt="Book cover"
          src="https://itbook.store/img/books/9781449331818.png"
        />
      </div>
      <div
        class="sc-dnqmqq bhqlKj"
      >
        <span
          class="sc-iwsKbI fJtlAM"
        >
          Learning JavaScript Design Patterns
        </span>
        <div />
        <div
          class="sc-gZMcBi cZDwvA"
        >
          $31.21
        </div>
        <button
          class="sc-bZQynM dlVnGl"
          type="button"
        >
          Add to cart
        </button>
      </div>
    </div>
  </li>
  <li>
    <div
      class="sc-gzVnrw hnuYoW"
    >
      <div
        class="sc-htoDjs cEeFSy"
      >
        <img
          alt="Book cover"
          src="https://itbook.store/img/books/9781449339951.png"
        />
      </div>
      <div
        class="sc-dnqmqq bhqlKj"
      >
        <span
          class="sc-iwsKbI fJtlAM"
        >
          The Problem with Native JavaScript APIs
        </span>
        <div />
        <div
          class="sc-gZMcBi cZDwvA"
        >
          $0.00
        </div>
        <button
          class="sc-bZQynM dlVnGl"
          type="button"
        >
          Add to cart
        </button>
      </div>
    </div>
  </li>
  <li>
    <div
      class="sc-gzVnrw hnuYoW"
    >
      <div
        class="sc-htoDjs cEeFSy"
      >
        <img
          alt="Book cover"
          src="https://itbook.store/img/books/9781449342883.png"
        />
      </div>
      <div
        class="sc-dnqmqq bhqlKj"
      >
        <span
          class="sc-iwsKbI fJtlAM"
        >
          JavaScript Enlightenment
        </span>
        <div />
        <div
          class="sc-gZMcBi cZDwvA"
        >
          $9.93
        </div>
        <button
          class="sc-bZQynM dlVnGl"
          type="button"
        >
          Add to cart
        </button>
      </div>
    </div>
  </li>
  <li>
    <div
      class="sc-gzVnrw hnuYoW"
    >
      <div
        class="sc-htoDjs cEeFSy"
      >
        <img
          alt="Book cover"
          src="https://itbook.store/img/books/9781449356378.png"
        />
      </div>
      <div
        class="sc-dnqmqq bhqlKj"
      >
        <span
          class="sc-iwsKbI fJtlAM"
        >
          JavaScript Testing with Jasmine
        </span>
        <div />
        <div
          class="sc-gZMcBi cZDwvA"
        >
          $10.25
        </div>
        <button
          class="sc-bZQynM dlVnGl"
          type="button"
        >
          Add to cart
        </button>
      </div>
    </div>
  </li>
</ul>
`;

exports[`App Router test should show ShoppingCartTable on /cart 1`] = `
<div
  class="sc-VigVT jShrPV"
  data-testid="test-cart-table"
>
  <h2>
    Your Order
  </h2>
  <table
    class="sc-jTzLTM dyJznJ"
  >
    <thead>
      <tr>
        <th>
          #
        </th>
        <th>
          Item
        </th>
        <th>
          Count
        </th>
        <th>
          Price
        </th>
        <th>
          Action
        </th>
      </tr>
    </thead>
    <tbody />
  </table>
  <div
    class="sc-fjdhpX cFfhmr"
  >
    Total: $
    0
  </div>
</div>
`;
